### ISSUE TITLE
encoding/*: allow exporting comments

### ISSUE THREAD

**Issue by romange** [cue-lang/cue#1180]
**Is your feature request related to a problem? Please describe.**
Sometimes 3rd party systems require annotating configurations via comments.
For example, cloud-init requires that its yaml files have a she-bang line like this:

```
#cloud-config
<some yaml stuff>
......
```
See here https://cloudinit.readthedocs.io/en/latest/topics/examples.html for more examples.

**Describe the solution you'd like**
File `test.cue` 
```
//cloud-config
foo: "bar"

hello: {
        // occupy
        world: "from mars?"
}
```
when exported via command `cue export --out yaml --preserve-comments test.cue` should produce something like this:
```
# cloud-config
foo: bar
hello:
  # occupy
  world: from mars?
```

**Describe alternatives you've considered**
There is an easy workaround when wrapping cue-cmd via bash script. It's suboptimal but it's not a blocker.

**Additional context**
Bonus point: Some comments may be relevant to cue language itself thus we do not really want to export them. 
It could be nice to have a convention to export whitelisted comments. For example, only comments starting from `///` will be exported.

**Comment by myitcv** [comment]
This request seems reasonable to me for one main reason: we import comments, hence we should export them for round-trip-ness' sake if nothing else.

FWIW it's not possible to control the export comments at all (unless I am missing something): neither through `cmd/cue`, `cue cmd`, nor `encoding/*`.

**Comment by verdverm** [comment]
@myitcv I agree that this makes a lot of sense for the round-trip-ness. I also did not see any way to export comments. My example above was a modified `cue` ( add `cue.Docs(true)` to this line https://github.com/cue-lang/cue/blob/master/pkg/encoding/yaml/manual.go#L38). 

Some thoughts after poking around

1. The example above adds a space within the comment string, I think we might consider removing the space here: https://github.com/cue-lang/cue/blob/master/internal/encoding/yaml/encode.go#L300 though i
1. Tt might be worth considering how we might preserve the spacing / padding in the original sources better. Does this need to happen for import as well?
1. My initial thought was to use the `Options` variadic func args, much like `Syntax` from the API since that is used where the change is needed. That being said, I'm not sure it makes sense to open up the `Options` that much for the encoders, as incorrect Options may cause errors in the encoder itself.
1. Does it make sense to allow different options depending on the output? I was mainly thinking that the CUE output could support the full set of Options, not sure if there is any overlap with CUE output and the feature request to support single file output for CUE we have seen. Protobuf looks to have some extra flags for exporting already.

**Comment by mpvl** [comment]
Makes a lot of sense. One alternative API is to export comments when used with `def`. I'm surprised that `cue def --out yaml` doesn't already export comments, actually.

**Comment by felixge** [comment]
I just ran into this problem and submitted a PR for it based on the approach suggested by @verdverm. I'm happy to adjust it if a different approach is preferred.

https://github.com/cue-lang/cue/pull/1549

**Comment by rogpeppe** [comment]
@felixge I was about to land your PR but first needed to fix a test case. In fixing that, I realised that perhaps this issue isn't as straightforward as it might seem.

Here's the test case in question: https://review.gerrithub.io/c/cue-lang/cue/+/536265/2/cmd/cue/cmd/testdata/script/cmd_github.txt

Note all the comments that have appeared in the YAML output for the test (a Github Actions file). Those comments aren't really relevant to the output file - they're really there to describe the _schema_ of the actions file, not the final rendered result. In the final rendered result, I'd imagine that more repository-specific comments would be appropriate.

Choosing a heuristic that decides which comments to be rendered in the final concrete output is tricky and will need some design input. Perhaps we might need to use an annotation hint; it could even be that it's not, after all, appropriate to include comments in rendered output.

So I've left the `NeedsDesign` label on this issue and removed the `GoodFirstIssue` label because it's clearly not entirely trivial!

Thanks for your input. Any further thoughts on this would be appreciated.

**Comment by felixge** [comment]
@rogpeppe I agree that this feature would benefit from a sophisticated design. But in the short term, I don't see the harm of letting the user chose to either include comments or not when they invoke the `cue` command (via a flag). It seems like most use cases would be well served by this?

**Comment by nyarly** [comment]
So long as we're in a `NeedsDesign` stage - got here because I was looking to produce comments based whether CUE processing of input data had fallen back on default values.

I realize that's a whole extra can of worms, but perhaps something like this could be made to work:
```cue
struct: {
  value: 17 @comment("from default")
}
```

**Comment by denniskern** [comment]
Hi Guys - Are there any ideas of how to proceed with this issue? It would be so nice to be able to export comments from a cue file.

**Comment by amackenzie** [comment]
This might be slightly out of scope, but...

To add a small comment, the OP brought up shebangs, and there's at least a few places where shebangs are used regardless of their validity in the normal export format. The one that comes to mind is Saltstack -- if you're using anything other than the Jinja|YAML renderer, you need a shebang at the top to designate it, eg, `#!json`, even though that isn't valid JSON. (And hypothetically you'd put a `#!cue` line at the top if there's a working cue renderer for Saltstack.)

So maybe in addition to comments there might need to be some consideration or syntax for fixed annotations of whatever format in outputs? It may affect importing if you want to be able to round-trip data, too, though at least in the specific case of shebangs they're generally the first line so it's skippable...

**Comment by DavidGamba** [comment]
@jpluscplusm described my need for round-tripping YAML comments very concisely here:

> please let me round-trip a given YAML file to CUE and back again, so that I can actually have *2* sources of truth, and give my 2 camps of users (team YAML vs team CUE) a way to work in their encoding of choice, without [bothering] the other set of users. /Whatever/ that implies, please just let my users get on with work, whilst I incrementally convince team YAML folks to defect to team CUE, and prevent pointless arguments about this stuff at the user-to-user level.

Source: https://cuelang.slack.com/archives/CLT4FD7KP/p1694007965956239

**Comment by jpluscplusm** [comment]
> @jpluscplusm described my need for round-tripping YAML comments very concisely here:
> 
> > please let me ...

For completeness, I will just mention that this paragraph was written and posted in the context of *me role-playing **as** David*, projecting my impression of their underlying, real-world requirements. I wasn't expressing what I needed, personally! :-)

**Comment by mvdan** [comment]
I began importing https://github.com/cue-lang/cue/pull/1549 into Gerrit in separate commits - first testdata files, then the Docs API fix that Felix spotted, and then after would come the enabling of comments when encoding YAML. See https://review.gerrithub.io/c/cue-lang/cue/+/1199690.

The Docs change uncovered a bit of a hidden problem: we never intended for `cue export` to output comments by default, [as can be seen in filetypes/types.cue](https://github.com/cue-lang/cue/blob/81d6f8bfcd61eeff2ec60ac71c8cc9867e6853e9/internal/filetypes/types.cue#L103). It just does for `cue export --out=cue` now due to the bug that Felix is fixing in the PR.

It's also reasonable for a user to want to either export with or without comments, so we should support both modes - this would require a new flag, at least to enable or disable the inclusion of doc comments. This would also require threading this option through the `internal/encoding` Go API as well as each of the encoding Go APIs, such as `encoding/yaml`.

Moreover, @rogpeppe and @mpvl argue that a boolean flag and option might not be enough. If CUE values unify concrete data with schemas, for example, they might both have doc comments. In some cases the user might want to output the comments from just the data, or from just the schema, or both combined. So perhaps a flag where one can do `--doc=all`, `--doc=schema`, `--doc=data`, `--doc=schema,data`, and so on would be better.

I will continue with this work step by step, but because we want to add the flag before "fixing" `cue export` to not output comments as intended, and multiple encodings are involved, this may take me some time.

I will use this issue to track the ability to enable exporting comments via `cmd/cue` and each of the encoding Go APIs. Some encodings simply don't support encoding or decoding comments yet, so that should be tracked separately, e.g. https://github.com/cue-lang/cue/issues/3342. This issue originally used YAML examples, so I'll aim to support comments in YAML as the first end-to-end example of this working.

**Comment by KeranYang** [comment]
Hey @mvdan , my team can really use this feature in our project. Do we expect this to be available any time soon? Thanks!

**Comment by mvdan** [comment]
@KeranYang after getting TOML to a working state and other work, this is back near the top of my TODO list. I can't promise any particular timeline but I hope to make progress here in the coming weeks.

**Comment by KeranYang** [comment]
Thank you very much @mvdan . Looking forward to it.

**Comment by rogpeppe** [comment]
> Choosing a heuristic that decides which comments to be rendered in the final concrete output is tricky and will need some design input.

I just had an idea about this.
- In general schemas will define optional and required fields, but not regular fields.
- Any comments on a schema will tend to be general comments about the field, not specific comments related to the actual configuration in question.
- The main reservation I had about this feature when I implemented it was that the comments were excessively noisy and hid all the actually relevant comments about the data in question; this was entirely due to the general schema comments.

So how about this: by default, data-oriented encodings should include comments from regular fields, but _not_ comments from optional or required fields. We could add a flag or encoding option to change this behavior to include all comments.

I _think_ this might end up with the right balance of signal to noise in exported comments.

**Comment by framegrace** [comment]
Support FLux ImageUpdate annotations may be the only thing stopping us from use cue. Having to post-process with sed or something like that just for this makes the whole system very fragile.
How about:
an attribute:
 @comment("{\"$imagepolicy\": \"flanks-data:crypter\"}")
Or a new token like:
` \\` instead of `//`  (Has a "Escaped comment" feeling)

**Comment by infogulch** [comment]
I guess the tension here is that the word "comment" actually has two different meanings in practice. If we define "code" as "the input of a translator", then "comment" can mean either:

1. Optional added text embedded alongside code to help other humans understand the code, but is ignored by translators
2. Actually it's just code

Good examples are Go's own comment directives like `//go:embed`, or the cloud config sigil mentioned in original issue post `#cloud-config`. Maybe shebangs count `#!/usr/bin/env bash`.

It seems to me that most of the time "comments interpreted as code" are intended to extend the syntax of the original language beyond what it is able to represent natively. Maybe it would behoove language designers to add extension points to begin with but it doesn't shake out that way very often in practice.

---

In any case this seems to be common enough that CUE can't really ignore this practice if it intends to serve as a universal translator for config languages; i.e. CUE should be able to serve definition 2 use-cases. But CUE users may also want to add regular comments in their CUE config; i.e. CUE should preserve the ability to serve definition 1 use-cases.

An `@comment` attribute may be a good way to go because it explicitly indicates that a comment should be emitted in the target document, but leaves regular cue comments alone.

**Comment by rogpeppe** [comment]
> I guess the tension here is that the word "comment" actually has two different meanings in practice. If we define "code" as "the input of a translator", then "comment" can mean either:
> 
> 1. Optional added text embedded alongside code to help other humans understand the code, but is ignored by translators
> 2. Actually it's just code

This observation is spot-on, thanks!

When I was talking about including comments [above](https://github.com/cue-lang/cue/issues/1180#issuecomment-2866821618), I was thinking of meaning 1, but @framegrace is clearly [talking about](https://github.com/cue-lang/cue/issues/1180#issuecomment-4287827936) meaning 2.

I tend to agree that in that case, an attribute might well be the correct approach.
However, I'm thinking that just a naked `@comment({"$imagepolicy": "flanks-data:crypter"})` attribute probably isn't desirable because that comment is clearly targetted at a particular output format, but might not be appropriate for all outputs.

I wonder about `@comment(someSpecifier, ...)` where `someSpecifier` is a token that signifies _when_ the comment should be included. For example, perhaps:

```
@comment(yaml, {"$imagepolicy": "flanks-data:crypter"})
```

So there the specifier specifies a target encoding file type, but maybe it's better to be some kind of tag that's specified explicitly on the command line, or... something else. Thoughts?

**Comment by infogulch** [comment]
Yes, there are a few important considerations regarding output comments:

1. What is the output file type (as Roger notes above)
2. Comment formatting
3. Comment type

Comment formatting can matter. For example, consider that a single-line comment in a JSON5 output format would need a newline after and maybe before the comment, but if there are no output comments then the whole document could be output in one big line with no linebreaks.

Comment type can be meaningful. For example, I don't think `//go:embed` works if it's a block comment like `/*go:embed*/`. Another example, [C# "documentation comments"](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/language-specification/documentation-comments) have a special comment sigils `///` or `/** **/` to denote text that should be included in the autogenerated documentation.

Some of these more complicated cases come from programming language comments which is not CUE's typical output file type, but maybe they should be considered anyway.

---

I like the idea of tagged `@comment` annotations, but it would be nice if there was a reasonable default. How about:

1. Add a `--doc` cli arg that enables all `@comment` annotations with matching tags. (Suggested by Daniel above.)
2. The `--out` cli arg automatically enables comments with the same tag, but you can override it by using the `--doc` arg and omitting the tag

So if you export with `--out yaml` then `@comment(yaml, "# yaml comment")` is enabled by default, but you can disable this comment with `--out yaml --doc ""`. Or you could have extra yaml comments like `@comment(myyaml, "# comment mccommentface")` that you can choose to enable by using `--out yaml --doc yaml,myyaml`.

---

As for the comment type problem, how about leaving the comment formatting completely up to the user? That is, anything in the second argument to the `@comment([tag], [comment text])` annotation should be inserted verbatim into the output.

The benefit is that this completely sidesteps the comment type problem and it gives users the ability to manage the comment formatting problem. The risk is that this puts all the complexity of comment handing onto the the CUE user and they could make a mistake (like: using `#` comment when the language uses `//` comments, forgetting the block comment closing `*/`, forgetting to add a newline after line comment, incorrect wrapping, etc). But these risks and complexity can be mitigated by writing cue expressions that help format comments as desired.
